web-console/src/lib/functions/ddl.ts

```diff
@@ -5,80 +5,110 @@
 // etc. This issue will go away once we support JSON for the data format. But
 // until then we do the type conversion here.
 
+import { clampBigNumber } from '$lib/functions/common/bigNumber'
 import { nonNull } from '$lib/functions/common/function'
+import { ColumnType, Relation, SqlType } from '$lib/services/manager'
 import assert from 'assert'
 import { BigNumber } from 'bignumber.js'
 import dayjs, { Dayjs, isDayjs } from 'dayjs'
 import invariant from 'tiny-invariant'
 import { match, P } from 'ts-pattern'
 
-import { GridCellParams } from '@mui/x-data-grid-pro'
+export type SQLValueJS = string | number | BigNumber | boolean | Dayjs | SQLValueJS[]
 
-import { ColumnType, Relation, SqlType } from '../services/manager'
-import { clampBigNumber } from './common/bigNumber'
-
-// Representing we get back from the ingress rest API.
+/**
+ * The format we get back from the ingress rest API.
+ */
 export interface Row {
   genId: number
   weight: number
-  record: Record<string, any>
+  record: Record<string, SQLValueJS | null>
 }
 
 export interface ValidationError {
   column: number
   message: string
 }
 
-// Returns a function which, for a given SQL type, converts the values of that
-// type to displayable strings.
-export function getValueFormatter(columntype: ColumnType): (value: any, params?: GridCellParams) => string {
-  return match(columntype)
+/**
+ * Returns a function which, for a given SQL type, converts the values of that
+ * type to displayable strings. null value for a nullable SQL type is rendered to null.
+ * @param columntype
+ * @returns
+ */
+export function getValueFormatter(columntype: ColumnType) {
+  const formatter = match(columntype)
+    .returnType<(value: SQLValueJS) => string>()
     .with({ type: SqlType.BIGINT }, { type: SqlType.DECIMAL }, () => {
-      return (value: BigNumber) => {
-        if (!BigNumber.isBigNumber(value)) {
-          return value
-        }
+      return value => {
+        invariant(BigNumber.isBigNumber(value))
         return value.toFixed()
       }
     })
+    .with(
+      { type: SqlType.TINYINT },
+      { type: SqlType.SMALLINT },
+      { type: SqlType.INTEGER },
+      { type: SqlType.REAL },
+      { type: SqlType.DOUBLE },
+      () => {
+        return value => {
+          invariant(typeof value === 'number' || BigNumber.isBigNumber(value))
+          return value.toString()
+        }
+      }
+    )
     .with({ type: SqlType.TIME }, () => {
-      return (value: Dayjs | string) => {
-        if (isDayjs(value)) {
-          return value.format('HH:mm:ss')
-        } else {
+      return value => {
+        invariant(typeof value === 'string' || isDayjs(value))
+        if (typeof value === 'string') {
           return value
         }
+        return value.format('HH:mm:ss')
       }
     })
     .with({ type: SqlType.DATE }, () => {
-      return (value: Dayjs | string) => {
+      return value => {
+        invariant(typeof value === 'string' && isDayjs(value))
```

Comment:
these invariants should not be checked here, but when you construct the value

Reply:
The code was initially loosely typed, so I am gradually tightening up the types and adding checks that I will be able to drop later on, as I do not want to spend a separate PR going over the values' lifecycle